fix(c/sql): jsqlparser for parsing SQL statements

zregvart opened the pull request:
The homegrown SQL parser made a number of assumptions and was very
brittle which made it really difficult to maintain.

This replaces the homegrown SQL parser with jsqlparser, which is more
robust, correct and feature full which should ease maintenance.

Ref. issues.redhat.com/browse/ENTESB-13689

fix(c/sql): make sure test tear down is done (6f85713)

We need to make sure that any object we create is dropped in the test.
Otherwise we might get flaky tests that depend on the order of
execution.

zregvart committed:
fix(c/sql): make sure test tear down is done

We need to make sure that any object we create is dropped in the test.
Otherwise we might get flaky tests that depend on the order of
execution.

zregvart committed:
fix(c/sql): jsqlparser for parsing SQL statements

The homegrown SQL parser made a number of assumptions and was very
brittle which made it really difficult to maintain.

This replaces the homegrown SQL parser with jsqlparser, which is more
robust, correct and feature full which should ease maintenance.

Ref. https://issues.redhat.com/browse/ENTESB-13689

KurtStam reviewed (commented):
I think it makes sense to open a separate issue for this as it is just the introduction of the jsqlparser. and do the UUID support in a different issue

zregvart commented:
> I think it makes sense to open a separate issue for this as it is just the introduction of the jsqlparser. and do the UUID support in a different issue

I've created the issue #9587, which is implemented by this pull request. The UUID support already went in #9248 and #9467. This pull request replaces the SQL parser with JSqlParser to enable the `CAST(x AS y)` syntax that would allow UUID column types to be used in expressions with PostgreSQL.

zregvart committed:
fix(c/sql): case insensitive missing table assert

It appears that some databases/JDBC drivers will return uppercased table
names. In the assertion that tries to see if the tables involved in the
statement exist in the schema/database we were case sensitive, this lead
to issues detected by the integration tests.

KurtStam reviewed (approved):
Thanks for doing this. Looks really good!

zregvart commented:
retest this please

mkralik3 commented:
retest this please

zregvart committed:
fix(c/sql): support SELECT without FROM

Some DBMS (PostgreSQL, SQLite) support `SELECT` without a `FROM`, in
that case we would run into an `ArrayIndexOutOfBoundsException` as we
would expect there to be at least one table involved in a `SELECT`
statement. Now we respect the possibility that there could be none.

zregvart commented:
retest this please

avano commented:
retest this please

zregvart commented:
What would you know, it's Tuesday already.
